### PR TITLE
+htp #16810 use flushing deflate call on Java > 6

### LINKED ISSUE #16810: For Deflate & Gzip compression use proper flushing APIs instead of hack
Currently we use this hack to ensure that the `java.util.zip.Deflater` flushes its buffers after each chunk:

```scala
  protected def flushWithBuffer(buffer: Array[Byte]): ByteString = {
    // trick the deflater into flushing: switch compression level
    // FIXME: use proper APIs and SYNC_FLUSH when Java 6 support is dropped
    deflater.deflate(EmptyByteArray, 0, 0)
    deflater.setLevel(Deflater.NO_COMPRESSION)
    val res1 = drain(buffer)
    deflater.setLevel(Deflater.BEST_COMPRESSION)
    val res2 = drain(buffer)
    res1 ++ res2
  }
```

Why this works is completely undocumented and e.g. doesn't work on Android.

Instead, the [deflate overload introduced in Java 7 that allows specifying a flush mode](https://docs.oracle.com/javase/7/docs/api/java/util/zip/Deflater.html#deflate(byte[],%20int,%20int,%20int)) should be used.

See spray/spray#1005.

### PR DESCRIPTION
**Pull request by jrudolph**
Fixes #16810.

/cc @sirthias

### CODE REVIEW

**Comment by jrudolph** [comment]
Failures in TlsSpec.

File: akka-http-core/src/main/scala/akka/http/impl/util/StreamUtils.scala [hunk 1]
```diff
@@ -27,16 +27,23 @@ private[http] object StreamUtils {
 
   /**
    * Creates a transformer that will call `f` for each incoming ByteString and output its result. After the complete
-   * input has been read it will call `finish` once to determine the final ByteString to post to the output.
+   * input has been read it will call `finish` once to determine the final ByteString to post to the output. Empty elements
+   * are filtered out.
```
Comment (viktorklang) [inline akka-http-core/src/main/scala/akka/http/impl/util/StreamUtils.scala]: May I suggest "Empty ByteStrings are discarded."

Comment (jrudolph) [inline akka-http-core/src/main/scala/akka/http/impl/util/StreamUtils.scala]: Yes, thanks. Updated.

**Comment by drewhk** [comment]
LGTM

**Comment by drewhk** [comment]
Fixes #16810

**Pull request merged**